### DISCUSSION TITLE
Jenkins pipeline setup

### DISCUSSION THREAD

**Discussion by abraluve** [nuke-build/nuke#1446] [Q&A] [answered]
Hi,

I am new to setting DevOps. I managed so far setup my few stages in Jenkins which works. Few weeks ago I heard about NUKE which is C# based, great. Unfortunately due to missing knowledge and failed to find in documentation I request your help.

Is NUKE meant to replace Jenkins pipeline syntax? So far I thought its possible to do all that steps based on few tutorials. However tutorails shows just basic stuff and they strictly rely on GitHub Actions.

My current setup is Jenkins which triggers pipeline by GitHub webhook on commit.

Steps:
```
pipeline {
    agent any
    environment {
        dotnet = 'C:\\Program Files\\dotnet\\dotnet.exe'
        BUILD_DIR = "C:\\Users\\my_user\\AppData\\Local\\Jenkins\\.jenkins\\workspace\\my_project\\src"
        PACKAGE_NAME = "my_build_name"
    }
    stages {
        stage('Checkout Stage') {
            steps {
                git credentialsId: 'xxx-xxx-xxx', url: 'https://github.com/xxx/xxx', branch: 'master'
            }
        }
        stage('Build Stage') {
            steps {
                bat 'dotnet build C:\\Users\\my_user\\AppData\\Local\\Jenkins\\.jenkins\\workspace\\my_project\\MyProject.sln --configuration Release'
            }
        }
        stage('Test Stage') {
            steps {
                bat 'dotnet test %WORKSPACE%\\Tests\\Tests.csproj'
            }
        }
        stage('Package') {
            steps {
                script {
                    bat '''
                        cd C:/jenkins/deploy
                        rmdir /s /q build
                        mkdir build
                        '''
                    bat """
                        dotnet publish %BUILD_DIR%\\MyProject.csproj -p:PublishProfile=%BUILD_DIR%\\Properties\\PublishProfiles\\FolderProfile.pubxml
                        """
                    PowerShell("Compress-Archive C:/jenkins/deploy/build C:/jenkins/deploy/${env.PACKAGE_NAME}.tmp.zip")
                    bat """
                        cd C:/jenkins/deploy
                        del /q %PACKAGE_NAME%.zip
                        ren %PACKAGE_NAME%.tmp.zip %PACKAGE_NAME%.zip
                        """
                }
            }
        }
        
        stage('Archive Artifacts') {
            steps {
                bat """
                    @echo off
                    if not exist \"%WORKSPACE%\\deploy\" (mkdir %WORKSPACE%\\deploy)
                    """
                bat "copy C:\\jenkins\\deploy\\%PACKAGE_NAME%.zip %WORKSPACE%\\deploy\\%PACKAGE_NAME%.zip"
                archiveArtifacts artifacts: "**/${env.PACKAGE_NAME}.zip", allowEmptyArchive: false
            }
        }
    }
}

def PowerShell(psCmd) {
    psCmd=psCmd.replaceAll("%", "%%")
    bat "powershell.exe -NonInteractive -ExecutionPolicy Bypass -Command \"\$ErrorActionPreference='Stop';[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;$psCmd;EXIT \$global:LastExitCode\""
}
```

Basically on webhook trigger I do repository checkout, then clean > build > package > archive. Right now I am not deploying anything, which I would like soon to get done.

Question is how to move this entire pipeline into NUKE. I would probably manage easier getting done Package and Archive Artifacts steps. Not quite sure how to make run NUKE build on GitHub webhook trigger in Jenkins. Certainly not sure how to do repository checkout using NUKE.

Can you guide me on this one?

**Comment by matkoch** [comment, marked as answer]
Nuke does not fully replace your `jenkinsfile`. Steps like checking out the repository still have to be defined as previously, but you get easier implementation for anything from calling .NET CLI, packaging artifacts, and deploying them somewhere. To invoke the build project, you just call `build.cmd` in your `jenkinsfile`.

**Comment by abraluve** [comment]
Thanks for reply.

I managed to figure it out as I found this article which really help me out. https://anktsrkr.github.io/post/write-your-first-building-block-in-nuke/

Now I know I can get rid of command like "dotnet build", replace my package stage etc. but still keep my Checkout stage.
